src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertyAndValueDeserializer.cs

```diff
@@ -206,28 +206,115 @@ internal object ReadODataOrCustomInstanceAnnotationValue(string annotationName,
         }
 
         /// <summary>
-        /// Check if a property value type in entity or complex value is deterministic .
+        /// Resolve the IEdmTypeReference of the current untyped value to be read.
         /// </summary>
         /// <param name="jsonReaderNodeType">The current JsonReader NodeType.</param>
         /// <param name="jsonReaderValue">The current JsonReader Value</param>
         /// <param name="payloadTypeName">The 'odata.type' annotation in payload.</param>
         /// <param name="payloadTypeReference">The payloadTypeReference of 'odata.type'.</param>
-        /// <returns>True if property value type is deterministic.</returns>
-        protected static bool IsKnownValueTypeForEntityOrComplex(JsonNodeType jsonReaderNodeType, object jsonReaderValue, string payloadTypeName, IEdmTypeReference payloadTypeReference)
+        /// <param name="primitiveTypeResolver">Function that takes a primitive value and returns an <see cref="IEdmTypeReference"/>.</param>
+        /// <param name="readUntypedAsString">Whether unknown properties should be read as a raw string value.</param>
+        /// <param name="generateTypeIfMissing">Whether to generate a type if not already part of the model.</param>
+        /// <returns>The <see cref="IEdmTypeReference"/> of the current value to be read.</returns>
+        [SuppressMessage("Microsoft.Performance", "CA1800:DoNotCastUnnecessarily")]
+        internal static IEdmTypeReference ResolveUntypedType(
+                JsonNodeType jsonReaderNodeType,
+                object jsonReaderValue,
+                string payloadTypeName,
+                IEdmTypeReference payloadTypeReference,
+                Func<object, string, IEdmTypeReference> primitiveTypeResolver,
+                bool readUntypedAsString,
+                bool generateTypeIfMissing)
         {
-            // Known
-            //   (1) recognized odata.type not Edm.Untyped
-            //   (2) no odata.type and value is bool
-            // Unknown
-            //    (1) unrecognized odata.type or 'Edm.Untyped'
-            //    (2) no odata.type and value is null / string / numeric / collection
-            if (string.IsNullOrEmpty(payloadTypeName))
+            if (payloadTypeReference != null && (payloadTypeReference.TypeKind() != EdmTypeKind.Untyped || readUntypedAsString))
             {
-                return (jsonReaderNodeType == JsonNodeType.PrimitiveValue) && (jsonReaderValue is bool);
+                return payloadTypeReference;
             }
-            else
+
+            if (readUntypedAsString)
             {
-                return (payloadTypeReference != null) && (payloadTypeReference.TypeKind() != EdmTypeKind.Untyped);
+                if (jsonReaderNodeType == JsonNodeType.PrimitiveValue && jsonReaderValue is bool)
+                {
+                    return EdmCoreModel.Instance.GetBoolean(true);

```

Comment:
Boolean values have a first class representation in JSON, so if the json value is a boolean we know it should be treated as such.

---
In reply to: [128108175](https://github.com/OData/odata.net/pull/894#discussion_r128108175) [](ancestors = 128108175)